### ISSUE TITLE
Error when snapshotting flutter_tool: Handle check failed: saw 291 expected AbstractType

### ISSUE THREAD

**Issue by jarekb123** [flutter/flutter#30102]
## Problem
Flutter upgrade crashes in master channel

## Steps to Reproduce

1. flutter channel master
2. flutter upgrade

## Logs
```
[  +16 ms] executing: [/home/xxx/sdks/flutter/] git log -n 1 --pretty=format:%H
[   +5 ms] Exit code 0 from: git log -n 1 --pretty=format:%H
[        ] 557c94ff778b94022adc3a8668e3d5706992aa1b
[        ] executing: [/home/xxx/sdks/flutter/] git describe --match v*.*.* --first-parent --long --tags
[   +8 ms] Exit code 0 from: git describe --match v*.*.* --first-parent --long --tags
[        ] v1.4.4-12-g557c94ff7
[   +3 ms] executing: [/home/xxx/sdks/flutter/] git rev-parse --abbrev-ref HEAD
[   +3 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] master
[   +2 ms] executing: [/home/xxx/sdks/flutter/] git describe --match v*.*.* --first-parent --long --tags
[   +9 ms] Exit code 0 from: git describe --match v*.*.* --first-parent --long --tags
[        ] v1.4.4-12-g557c94ff7
[   +2 ms] executing: [/home/xxx/sdks/flutter/] git rev-parse @{u}
[  +14 ms] 557c94ff778b94022adc3a8668e3d5706992aa1b
[   +1 ms] executing: [/home/xxx/sdks/flutter/] git reset --hard v1.4.4
[  +43 ms] HEAD is now at 3b3f6c7a0 Include platformViewId in semantics tree for iOS (#29304)
[   +1 ms] Upgrading Flutter from /home/xxx/sdks/flutter...
[   +2 ms] executing: [/home/xxx/sdks/flutter/] git rev-parse --abbrev-ref --symbolic @{u}
[   +3 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[        ] origin/master
[        ] executing: [/home/xxx/sdks/flutter/] git ls-remote --get-url origin
[   +3 ms] Exit code 0 from: git ls-remote --get-url origin
[        ] https://github.com/flutter/flutter.git
[   +2 ms] executing: [/home/xxx/sdks/flutter/] git pull --ff
[+1568 ms] Updating 3b3f6c7a0..557c94ff7
[  +13 ms]  76 files changed, 607 insertions(+), 546 deletions(-)
[   +1 ms] Upgrading engine...
[        ] executing: [/home/xxx/sdks/flutter/] bin/flutter --no-color --no-version-check precache
[  +34 ms] Building flutter tool...
../../third_party/dart/runtime/vm/object.h: 6486: error: Handle check failed: saw 291 expected AbstractType
version=2.2.1-dev.2.0.flutter-None (Wed Mar 27 14:11:17 2019 +0000) on "linux_x64"
thread=4682, isolate=flutter_tools.snapshot:main()(0x21a1df0)
  [0x00000000014ec0fc] Unknown symbol
  [0x00000000014ec0fc] Unknown symbol
-- End of DumpStackTrace
/home/xxx/sdks/flutter/bin/flutter: line 177:  4546 Aborted                 (core dumped) "$DART" $FLUTTER_TOOL_ARGS "$SNAPSHOT_PATH" "$@"
```

Flutter doctor
```
[✓] Flutter (Channel master, v1.4.5-pre.12, on Linux, locale en_US.UTF-8)
    • Flutter version 1.4.5-pre.12 at /home/jbutajlo/sdks/flutter
    • Framework revision 557c94ff77 (3 hours ago), 2019-03-28 01:00:29 -0700
    • Engine revision 37947f961a
    • Dart version 2.2.1 (build 2.2.1-dev.2.0 None)

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /home/jbutajlo/Android/Sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: /snap/android-studio/73/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)
    • All Android licenses accepted.

[✓] Android Studio (version 3.3)
    • Android Studio at /snap/android-studio/73/android-studio
    • Flutter plugin version 33.4.1
    • Dart plugin version 182.5215
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)

[✓] IntelliJ IDEA Community Edition (version 2018.3)
    • IntelliJ at /snap/intellij-idea-community/126
    • Flutter plugin version 33.4.2
    • Dart plugin version 183.5912.23

[!] Connected device
    ! No devices available

! Doctor found issues in 1 category.
```

**Comment by mraleph** [comment]
If you can reliably reproduce this please enable core dumps (e.g. `ulimit -c unlimited`), reproduce the issue and upload produced core file.

**Comment by jason-simmons** [comment]
I can reproduce this by downloading the Flutter v1.2.1 release tarball from https://flutter.dev/docs/get-started/install/linux, running `flutter channel master`, and then running `flutter upgrade`

Core dump attached.

Output from `flutter upgrade`:
```
Upgrading engine...
Building flutter tool...

===== CRASH =====
si_signo=Bus error(7), si_code=2, si_addr=0x7f755bc84620
version=2.2.1-dev.2.0.flutter-None (Thu Mar 28 00:48:30 2019 +0000) on "linux_x64"
thread=166629, isolate=flutter_tools.snapshot:main()(0x34a9480)
  [0x00007f755bc84620] Unknown symbol
  [0x00007f755bc84620] Unknown symbol
-- End of DumpStackTrace
bin/flutter: line 177: 166448 Aborted                 "$DART" $FLUTTER_TOOL_ARGS "$SNAPSHOT_PATH" "$@"
```

[bug_30102_core.gz](https://github.com/flutter/flutter/files/3023930/bug_30102_core.gz)

**Comment by jason-simmons** [comment]
For now we're working around this by disabling app-jit snapshots.

See https://github.com/flutter/flutter/pull/30204

**Comment by jonahwilliams** [comment]
This was fixed by #30204
